src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Internal/MsQuicApi.cs

```diff
@@ -116,6 +120,16 @@ static MsQuicApi()
                 return;
             }
 
+            if (isWindows && IsTls1_3Disabled())
+            {
+                if (NetEventSource.Log.IsEnabled())
+                {
+                    NetEventSource.Info(null, $"TLS 1.3 has been disabled via registry. TLS 1.3 is required for QUIC.");
```

Comment:
I'm not sure if this saves the problem. I assume the @JamesNK  meant QuicException.Message when he talk about "message" e.g. surface that info to the caller.

Reply:
This is "option 1.", i.e.: including TLS 1.3 check within the `IsSupported` flag, so if TLS 1.3 is off, Quic won't light up. Whereas previously it would light up and then throw `0x80090331`.
We might opt to throw a better error, e.g. special case for `0x80090331`, translate it to "Algorithm mismatch" and add text to the exception saying: "Make sure you have TLS 1.3 enabled: link to KB or something"

Reply:
Yeah, I was aiming for the "option 1" route originally. Although I think given that Client and Server can be independently disabled, it could be problematic/misleading to include TLS being disabled in setting `IsQuicSupported`.

I think the `Tls13MayBeDisabled` and special case error handling mentioned by wfurt below makes sense, as it allows for Quic to be used for a client use case when the TLS 1.3 server role has been disabled.

Reply:
fine with me. That path creates silent failure that will be harder to diagnose IMHO. But that scenario may be rare and it still may be better than failing operations.

Reply:
Other than that, a simpler and more catch all option would be just to special case `0x80090331` in `QuicExceptionHelper` and have it add information relevant to the TLS 1.3 registry settings?

I think that would probably give better coverage of users that disable TLS after the launch, my only worry would be the potential to mask or mislead in cases where disabled TLS isn't the cause of the `0x80090331`

Reply:
Would it be valid to move the registry check to the point at which the exception is being thrown? `QuicExceptionHelpers.CreateExceptionForHResult()`, that way we get a more accurate "point-in-time" representation of whether TLS 1.3 is currently disabled, but we're still only making Registry queries in the case of a `0x80090331`.

Something like: 

```c#
        internal static Exception CreateExceptionForHResult(uint status, string? message = null, Exception? innerException = null)
        {
            if (status == 0x80090331 && IsTlsDisabledInTheRegistry())
            {
                return new QuicTlsDisabledException();
            }
            else
            {
                return new QuicException($"{message} Error Code: {MsQuicStatusCodes.GetError(status)}", innerException, MapMsQuicStatusToHResult(status));
            }
        }
```